packages/flutter/lib/src/material/page_transitions_theme.dart

```diff
@@ -0,0 +1,352 @@
+                      animation: secondaryAnimation,
+                      child: FractionalTranslation(
+                        translation: primaryTranslationAnimation.value,
+                        child: child,
+                      ),
+                      builder: (BuildContext context, Widget child) {
+                        return FractionalTranslation(
+                          translation: secondaryTranslationAnimation.value,
+                          child: child,
+                        );
+                      },
+                    ),
+                  ),
+                ),
+              ),
+            );
+          },
+        );
+      },
+    );
+  }
+}
+
+/// Used by [PageTransitionsTheme] to define a [MaterialPageRoute] page
+/// transition animation.
+///
+/// If [platform] is non-null then this transition is preferred when
+/// it matches the current platform, [ThemeData.platform].
+///
+/// Apps can configure the list of builders for [ThemeData.platformTheme]
+/// to customize the default [MaterialPageRoute] page transition animation
+/// for different platforms.
+///
+/// See also:
+///
+///  * [GenericPageTransitionsBuilder], which defines a default page transition.
+///  * [MountainViewPageTransitionsBuilder], which defines a page transition
+///    that's similar to the one provided by Android P.
+///  * [CupertinoPageTransitionsBuilder], which defines a horizontal page
+///    transition that matches nativer iOS page transitions.
+abstract class PageTransitionsBuilder {
+  const PageTransitionsBuilder({ this.platform });
+
+  final TargetPlatform platform;
+
+  Widget buildTransitions<T>(
+    PageRoute<T> route,
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  );
+
+  @override
+  String toString() {
+    return '[$runtimeType${platform == null ? "" : " for $platform"}]';
+  }
+}
+
+/// Used by [PageTransitionsTheme] to define a default [MaterialPageRoute] page
+/// transition animation.
+///
+/// The default animation fades the new page in while translating it upwards,
+/// starting from about 25% below the top of the screen.
+///
+/// The [platform] for this builder is null which indicates that it's not
+/// platform-specific.
+///
+/// See also:
+///
+///  * [MountainViewPageTransitionsBuilder], which defines a page transition
+///    that's similar to the one provided by Android P.
+///  * [CupertinoPageTransitionsBuilder], which defines a horizontal page
+///    transition that matches nativer iOS page transitions.
+class GenericPageTransitionsBuilder extends PageTransitionsBuilder {
+  const GenericPageTransitionsBuilder() : super(platform: null);
+
+  @override
+  Widget buildTransitions<T>(
+    PageRoute<T> route,
+    BuildContext context,
+    Animation<double> animation,
+    Animation<double> secondaryAnimation,
+    Widget child,
+  ) {
+    return _GenericPageTransition(routeAnimation: animation, child: child);
```

Comment:
is there much benefit to having the _GenericPageTransition widget rather than just inlining its build method here?

Reply:
I don't see how it helps with this? The widgets are private, so the effect seems to be identical, from a black-box perspective, as having the widgets inlined.